### ISSUE TITLE
How to process the 'ZeroDivisionError: matrix is numerically singular'

### ISSUE THREAD

**Issue by DavidLining** [sympy/sympy#14426]
I try to use scipy and sympy to solve same equations. But encounter the 'ZeroDivisionError: matrix is numerically singular' when i using lib sympy.
Following are codes:

def solve_eq_scipy(x):
    x1, y1, z1, x2, y2, z2, x3, y3, z3 = x.tolist()
    eqs = []
    f1 = x1**2 + y1**2 + z1**2 - 20.327 #red
    f2 = 487.5*x1 - 1398.5*y1
    f3 = x2**2 + y2**2 + z2**2 - 22.325 #blue
    f4 = 403.5*x2 - 1482.5*y2
    f5 = x3**2 + y3**2 + z3**2 - 26.329 #green
    f6 = 570.5*x3 - 1650.5*y3
    f7 = (x1 - x2)**2 + (y1- y2)**2 + (z1 - z2)**2 -0.25
    f8 = (x1 - x3)**2 + (y1- y3)**2 + (z1 - z3)**2 -1.25
    f9 = (x3 - x2)**2 + (y3- y2)**2 + (z3 - z2)**2 -1.0 
    return [f1, f2, f3, f4, f5, f6, f7, f8, f9]    
    
def solve_eq_sympy():
    x1, y1, z1, x2, y2, z2, x3, y3, z3 = symbols('x1 y1 z1 x2 y2 z2 x3 y3 z3')
    eqs = []
    f1 = x1**2 + y1**2 + z1**2 - 20.327 #red
    f2 = 487.5*x1 - 1398.5*y1
    f3 = x2**2 + y2**2 + z2**2 - 22.325 #blue
    f4 = 403.5*x2 - 1482.5*y2
    f5 = x3**2 + y3**2 + z3**2 - 26.329 #green
    f6 = 570.5*x3 - 1650.5*y3
    f7 = (x1 - x2)**2 + (y1- y2)**2 + (z1 - z2)**2 -0.25
    f8 = (x1 - x3)**2 + (y1- y3)**2 + (z1 - z3)**2 -1.25
    f9 = (x3 - x2)**2 + (y3- y2)**2 + (z3 - z2)**2 -1.0 
    print(sympy.nsolve((f1, f2, f3, f4, f5, f6, f7, f8, f9), \
                 (x1, y1, z1, x2, y2, z2, x3, y3, z3), 
                 (1.0, 1.0 ,1.0 ,1.0, 1.0, 1.0, 1.0, 1.0, 1.0)))    

result = fsolve(solve_eq_scipy, [4.25, 1.0 ,1.0 ,1.0, 1.0, 1.0, 1.0, 1.0, 1.0])
print(result)
print(solve_eq_scipy(result))
print(solve_eq_sympy())

I don't know what happened in above problem? It's sames that scipy can solve it.

**Comment by DavidLining** [comment]
[def solve_eq_scipy(x):
    x1, y1, z1, x2, y2, z2, x3, y3, z3 = x.tolist()
    eqs = []
    f1 = (x1)**2 + (y1)**2 + (z1)**2 - 20.327 #red
    f2 = 487.5*(x1) - 1398.5*(y1)
    f3 = (x2)**2 + (y2)**2 + (z2)**2 - 22.325 #blue
    f4 = 403.5*(x2) - 1482.5*(y2)
    f5 = (x3)**2 + (y3)**2 + (z3)**2 - 26.329 #green
    f6 = 570.5*(x3) - 1650.5*(y3)
    f7 = (x1 - x2)**2 + (y1- y2)**2 + (z1 - z2)**2 -0.25
    f8 = (x1 - x3)**2 + (y1- y3)**2 + (z1 - z3)**2 -1.25
    f9 = (x3 - x2)**2 + (y3- y2)**2 + (z3 - z2)**2 -1.0 
    return [f1, f2, f3, f4, f5, f6, f7, f8, f9]    
    
def solve_eq_sympy():
    x1, y1, z1, x2, y2, z2, x3, y3, z3 = symbols('x1 y1 z1 x2 y2 z2 x3 y3 z3')
    eqs = []
    f1 = (x1)**2 + (y1)**2 + (z1)**2 - 20.327 #red
    f2 = 487.5*(x1) - 1398.5*(y1)
    f3 = (x2)**2 + (y2)**2 + (z2)**2 - 22.325 #blue
    f4 = 403.5*(x2) - 1482.5*(y2)
    f5 = (x3)**2 + (y3)**2 + (z3)**2 - 26.329 #green
    f6 = 570.5*(x3) - 1650.5*(y3)
    f7 = (x1 - x2)**2 + (y1- y2)**2 + (z1 - z2)**2 -0.25
    f8 = (x1 - x3)**2 + (y1- y3)**2 + (z1 - z3)**2 -1.25
    f9 = (x3 - x2)**2 + (y3- y2)**2 + (z3 - z2)**2 -1.0 
    print(sympy.nsolve((f1, f2, f3, f4, f5, f6, f7, f8, f9), \
                 (x1, y1, z1, x2, y2, z2, x3, y3, z3), 
                 (1.0, 1.0 ,1.0 ,1.0, 1.0, 1.0, 1.0, 1.0, 1.0)))    

result = fsolve(solve_eq_scipy, [4.25, 1.0 ,1.0 ,1.0, 1.0, 1.0, 1.0, 1.0, 1.0])
print(result)
print(solve_eq_scipy(result))
print(solve_eq_sympy())](url)

**Comment by DavidLining** [comment]
Code:
[solve.txt](https://github.com/sympy/sympy/files/1792426/solve.txt)

**Comment by normalhuman** [comment]
Both SciPy and SymPy have difficulties finding a solution here (if it even exists). Two issues:

Poor choice of the starting point. Giving equal values to all coordinates makes the gradient of the terms like `(x1 - x2)**2` equal to zero, and zero gradient is a problem for solvers (degenerate Jacobian matrix is the primary source for division by zero errors in this process). I notice that the SciPy version has one component not equal to others. But to SymPy you give all equal coordinates. Use another starting point, like
```
(4.25, 4.0, 3.5, 3.0, 2.7, 2.3, 2.0, 1.5, 1.0)
```

Different criteria for returning a value. SciPy issues a warning about poor progress, but still returns a "solution" with some residuals of the size about 0.1, leaving it for you to decide if you like it (I'm not sure I do). To mimic this behavior in SymPy, pass `verify=False` to `nsolve`. 
```
sympy.nsolve((f1, f2, f3, f4, f5, f6, f7, f8, f9),
                 (x1, y1, z1, x2, y2, z2, x3, y3, z3), 
                 (4.25, 4.0, 3.5, 3.0, 2.7, 2.3, 2.0, 1.5, 1.0),
                      verify=False)) 
```
returns 
```
Matrix([[10.0054882701604], [3.48779087000585], [2.46630991423488], [7.85057136301067], [2.13673223944338], [-0.212558387367815], [5.92314641332844], [2.04735233493116], [-2.87288229041487]])
```
I haven't checked how good this result is; chances are there are some nontrivial residuals there. SciPy with the same initial point returns something with much larger residuals. If the initial point is changed a bit, the output changes by quite a lot. 
 
I don't know if the system even has a solution at all; if it doesn't, that would explain why both solvers struggle with it. 

In any case, there is no particular reason to use SymPy here. It does not have a numeric solver; it just passes the problem to [mpmath.findroot](http://mpmath.org/doc/1.0.0/calculus/optimization.html). The documentation of findroot offers some choice of solution methods and parameters to use. 

So do SciPy solvers, of course.

**Comment by DavidLining** [comment]
@normalhuman Thanks a lot, it sames that Scipy is much better than Sympy on solving nolinear equations. 
Sympy is outstanding on symbolic process. And that's why I want to use Sympy here. In my project, it may be have thousands of equations and  unknowns to solve. It sames that I need to write all of equations  if i using Scipy. However, I can use looping method to define variables and equations with Sympy.

**Comment by DavidLining** [comment]
@normalhuman  Is there anyway to use Sympy symbolic express such as 'symbols, Eq()' with Scipy function such as 'root, fsolve' ?

**Comment by normalhuman** [comment]
No. SymPy objects are understood only by SymPy. But there is `lambdify` which can make a Python function out of a SymPy expression.

>  I need to write all of equations if i using Scipy. However, I can use looping method to define variables and equations with Sympy

This is not convincing. Loops work with SciPy, too. Specifically, the objective function receives a NumPy array and can use whatever loops are necessary to process that array and return another array (the left hand sides of the equations in the system).
